### PR TITLE
Optimize bpf only

### LINKED ISSUE #10: Optimize marine_inner_dissect_packet when only BPF is used
If only BPF is used, we don't need to continue execution after filtering with BPF.

This will improve speeds and memory usage significantly in this case.

### PR DESCRIPTION
**Pull request by royreznik**
resolve #10

### CODE REVIEW

File: marine_benchmark.c [hunk 1]
```diff
@@ -11,10 +11,10 @@
 #define ARRAY_SIZE(arr)     (sizeof(arr) / sizeof((arr)[0]))
 
 typedef struct {
-    char* title;
-    char* bpf;
-    char* dfilter;
-    char** fields;
+    char *title;
+    char *bpf;
+    char *dfilter;
```
Comment (yehonatanz) [inline marine_benchmark.c:16]: Awesome

File: marine_benchmark.c [hunk 2]
```diff
@@ -77,26 +77,28 @@ int load_cap(char *file, packet **packets, char errbuff[PCAP_ERRBUF_SIZE]) {
...
     }
-    clock_t end = clock();
+    clock_gettime(CLOCK_MONOTONIC_RAW, &end_time);
     size_t memory_end = get_current_rss();
 
-    float total_time = ((float) end - start) / CLOCKS_PER_SEC;
-    float pps = (float) packet_len / total_time;
-    float memory_usage = ((float) memory_end - memory_start) / 1024 / 1024;
+    double total_time = (end_time.tv_sec - start_time.tv_sec) + ((end_time.tv_nsec - start_time.tv_nsec) * 1e-9);
+    double pps = packet_len / ((double) total_time);
```
Comment (yehonatanz) [inline marine_benchmark.c]: `total_time` is already double, why are you casting it?

File: marine.c [hunk 3]
```diff
@@ -230,6 +230,9 @@ static gboolean prefs_loaded = FALSE;
 
 static void reset_epan_mem(capture_file *cf, epan_dissect_t *edt, gboolean tree, gboolean visual);
 
+inline static int is_only_bpf(packet_filter* filter) {
```
Comment (yehonatanz) [inline marine.c]: Your don't modify your pointer or its data, use `const packet_filter* const filter`.

**Review by yehonatanz — CHANGES_REQUESTED** [review]

File: marine_benchmark.c (hunk 2 above)
Comment (royreznik) [reply marine_benchmark.c]: idk

**Review by yehonatanz — APPROVED** [review]
Approve, but for future reviewers, please comment with the BPF PPS before and after your change

File: marine.c [hunk 4]
```diff
@@ -507,6 +510,9 @@ marine_inner_dissect_packet(capture_file *cf, packet_filter *filter, const unsig
             return 0;
         }
         free(hdr);
+        if (is_only_bpf(filter)) {
+            return 1;
```
Comment (tomlegkov) [inline marine.c]: `return TRUE;`

File: marine.c [hunk 5]
```diff
@@ -230,6 +230,9 @@ static gboolean prefs_loaded = FALSE;
 
 static void reset_epan_mem(capture_file *cf, epan_dissect_t *edt, gboolean tree, gboolean visual);
 
+inline static int is_only_bpf(const packet_filter* const filter) {
```
Comment (tomlegkov) [inline marine.c:233]: Why is there const twice here?

**Review by tomlegkov — CHANGES_REQUESTED** [review]

File: marine.c (hunk 5 above)
Comment (royreznik) [reply marine.c:233]: Const pointer, of const value

**Review by tomlegkov — APPROVED** [review]

**Pull request merged**